File: http/http-server-annotation-processor/src/test/java/ru/tinkoff/kora/http/server/annotation/processor/ExtensionTest.java
package ru.tinkoff.kora.http.server.annotation.processor;

import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.annotation.processor.common.AbstractAnnotationProcessorTest;
import ru.tinkoff.kora.kora.app.annotation.processor.KoraAppProcessor;

import java.util.List;

public class ExtensionTest extends AbstractAnnotationProcessorTest {
    @Override
    protected String commonImports() {
        return super.commonImports() + """
            import ru.tinkoff.kora.http.server.common.handler.*;
            import ru.tinkoff.kora.http.server.common.*;
            import ru.tinkoff.kora.http.common.*;
            """;
    }

    @Test
    public void testExtensionWithTag() {
        compile(List.of(new KoraAppProcessor()), """
            @KoraApp
            public interface Controller {
                @Root
                default String root(@Tag(String.class) HttpServerResponseMapper<HttpResponseEntity<String>> mapper) { return ""; }
            
                @Tag(String.class)
                default HttpServerResponseMapper<String> mapper() { return (rq, result) -> HttpServerResponse.of(200); }
            }
            """);

        compileResult.assertSuccess();
    }

    @Test
    public void testExtensionWithoutTag() {
        compile(List.of(new KoraAppProcessor()), """
            @KoraApp
            public interface Controller {
                @Root
                default String root(HttpServerResponseMapper<HttpResponseEntity<String>> mapper) { return ""; }
            
                default HttpServerResponseMapper<String> mapper() { return (rq, result) -> HttpServerResponse.of(200); }
            }
            """);

        compileResult.assertSuccess();
    }
}

File: http/http-server-symbol-processor/src/test/kotlin/ru/tinkoff/kora/http/server/symbol/processor/ExtensionTest.kt
package ru.tinkoff.kora.http.server.symbol.processor

import org.junit.jupiter.api.Test
import ru.tinkoff.kora.kora.app.ksp.KoraAppProcessorProvider
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest


class ExtensionTest : AbstractSymbolProcessorTest() {
    override fun commonImports(): String {
        return super.commonImports() + """
            import ru.tinkoff.kora.http.server.common.handler.*
            import ru.tinkoff.kora.http.server.common.*
            import ru.tinkoff.kora.http.common.*
            """.trimIndent()
    }

    @Test
    fun testExtensionWithTag() {
        compile0(
            listOf(KoraAppProcessorProvider()), """
            @KoraApp
            interface Controller {
                @Root
                fun root(@Tag(String::class) mapper: HttpServerResponseMapper<HttpResponseEntity<String>>): String = ""
            
                @Tag(String::class)
                fun mapper() : HttpServerResponseMapper<String> = HttpServerResponseMapper<String> { rq, result  -> HttpServerResponse.of(200) }
            }
            """.trimIndent()
        )

        compileResult.assertSuccess()
    }

    @Test
    fun testExtensionWithoutTag() {
        compile0(
            listOf(KoraAppProcessorProvider()), """
            @KoraApp
            interface Controller {
                @Root
                fun root(mapper: HttpServerResponseMapper<HttpResponseEntity<String>>): String = ""
            
                fun mapper() : HttpServerResponseMapper<String> = HttpServerResponseMapper<String> { rq, result  -> HttpServerResponse.of(200) }
            }
            """.trimIndent()
        )

        compileResult.assertSuccess()
    }

    @Test
    fun testExtensionWithNullable() {
        compile0(
            listOf(KoraAppProcessorProvider()), """
            @KoraApp
            interface Controller {
                @Root
                fun root(mapper: HttpServerResponseMapper<HttpResponseEntity<String?>>): String = ""
            
                fun mapper() : HttpServerResponseMapper<String> = HttpServerResponseMapper<String> { rq, result  -> HttpServerResponse.of(200) }
            }
            """.trimIndent()
        )

        compileResult.assertSuccess()
    }
}
